src/test/java/plugins/GitPluginTest.java

```diff
@@ -209,16 +202,226 @@ public void update_submodules_recursively() throws IOException, InterruptedExcep
+                .sparseCheckout().addPath(SUB_DIR);
+
+        job.addShellStep("test ! -f " + TEST_FILE + " && test -f " + SUB_DIR + "/" + TEST_FILE);
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+    }
+
+    @Test
+    public void ancestry_strategy_to_choose_build() {
+        final String TEST_BRANCH = "testBranch";
+
+        GitRepo repo = buildGitRepo();
+        repo.createBranch(TEST_BRANCH);
+        repo.changeAndCommitFoo("Commit1 on master");
+        repo.checkout(TEST_BRANCH);
+        repo.changeAndCommitFoo("commit1 on " + TEST_BRANCH);
+        String sha1 = repo.getLastSha1();
+        repo.changeAndCommitFoo("commit2 on " + TEST_BRANCH);
+
+        repo.transferToDockerContainer(host, port);
+
+        job.useScm(GitScm.class)
+                .url(repoUrl)
+                .branch("")
+                .credentials(USERNAME)
+                .chooseBuildStrategy("Ancestry", 1, sha1);
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+        // TODO Multiple selected branches create multiple builds, these should also be verified
+
+        assertThat(
+                b.getConsole(),
+                Matchers.containsRegexp(
+                        "Checking out Revision .* \\(origin/"+TEST_BRANCH+"\\)",
+                        Pattern.MULTILINE
+                )
+        );
+    }
+
+    @Test
+    public void inverse_strategy_to_choose_build() {
+        final String BRANCH_NAME = "secondBranch";
+
+        GitRepo repo = buildGitRepo();
+        repo.createBranch(BRANCH_NAME);
+        repo.transferToDockerContainer(host, port);
+
+        job.useScm(GitScm.class)
+                .url(repoUrl)
+                .credentials(USERNAME)
+                .chooseBuildStrategy("Inverse");
+
+        job.save();
+        Build b = job.startBuild();
+        b.shouldSucceed();
+
+        assertThat(
+                b.getConsole(),
+                Matchers.containsRegexp(
+                        "Checking out Revision .* \\(origin/"+BRANCH_NAME+"\\)",
+                        Pattern.MULTILINE
+                )
+        );
+    }
+
+    ////////////////////
+    // HELPER METHODS //
+    ////////////////////
+
     private String getRevisionFromConsole(String console) {
         Pattern p = Pattern.compile("(?<=\\bRevision\\s)(\\w+)");
         Matcher m = p.matcher(console);
         assertThat(m.find(), is(true));
         return m.group(0);
     }
 
-    private GitRepo buildGitRepo() throws IOException, InterruptedException {
+    private GitRepo buildGitRepo() {
         GitRepo repo = new GitRepo();
         repo.changeAndCommitFoo("Initial commit");
         return repo;
     }
+
+    /**
+     * Invoked by {@link #clean_after_checkout()} and {@link #clean_before_checkout()}
+     *
+     * @param before Select "clean before" or "clean after"
+     */
+    private void test_clean_while_checkout(boolean before) {
```

Comment:
@MarkEWaite what is the difference between clean before/after checkout? Is there an easy way to verify the differences of these two variants?